src/target/stm32h7.c

```diff
@@ -343,25 +371,56 @@ static bool stm32h7_flash_unlock(target_s *const target, const uint32_t addr)
 	return !(target_mem_read32(target, regbase + FLASH_CR) & FLASH_CR_LOCK);
 }
 
+/* Helper for offsetting FLASH_CR bits correctly */
+static uint32_t stm32h7_flash_cr(uint32_t sector_size, const uint32_t ctrl, unsigned int snb)
+{
+	uint32_t command = ctrl;
+	/* H74x, H72x IP: 128 KiB and has PSIZE */
+	if (sector_size == FLASH_SECTOR_SIZE) {
+		command |= snb * FLASH_CR_SNB_1;
+		DEBUG_TARGET("%s: patching FLASH_CR from 0x%08" PRIx32 " to 0x%08" PRIx32 "\n", __func__, ctrl, command);
+		return command;
+	}
+
+	/* H7Bx IP: 8 KiB and no PSIZE */
+	/* Save and right-shift FW, START bits */
+	const uint32_t temp_fw_start = command & (FLASH_CR_FW | FLASH_CR_START);
+	/* Parallelism is ignored */
+	command &= ~(FLASH_CR_PSIZE64 | FLASH_CR_FW | FLASH_CR_START);
+	/* Restore FW, START to H7Bx-correct bits */
+	command |= temp_fw_start >> 2U;
+	/* SNB offset is different, too */
+	command |= snb << 6U;
+	DEBUG_TARGET("%s: patching FLASH_CR from 0x%08" PRIx32 " to 0x%08" PRIx32 "\n", __func__, ctrl, command);
+	return command;
+}
+
 static bool stm32h7_flash_erase(target_flash_s *const target_flash, target_addr_t addr, const size_t len)
 {
+	const uint32_t sector_size = target_flash->blocksize;
 	target_s *target = target_flash->t;
 	const stm32h7_flash_s *const flash = (stm32h7_flash_s *)target_flash;
 	/* Unlock the Flash */
 	if (!stm32h7_flash_unlock(target, addr))
 		return false;
 	/* We come out of reset with HSI 64 MHz. Adapt FLASH_ACR.*/
-	target_mem_write32(target, flash->regbase + FLASH_ACR, 0);
-	addr &= (NUM_SECTOR_PER_BANK * FLASH_SECTOR_SIZE) - 1U;
-	const size_t end_sector = (addr + len - 1U) / FLASH_SECTOR_SIZE;
+	uint32_t acr = 0;
+	/* H7Bx starts up in VOS3 from HSI 64 MHz with default ACR of 0x13 */
+	if (sector_size == 0x2000U)
+		acr = 0x13;
+	target_mem_write32(target, flash->regbase + FLASH_ACR, acr);
+	/* Calculate SNB span */
+	addr &= target_flash->length - 1U;
+	const size_t end_sector = (addr + len - 1U) / sector_size;
 	const align_e psize = flash->psize;
 	const uint32_t reg_base = flash->regbase;
 
-	for (size_t begin_sector = addr / FLASH_SECTOR_SIZE; begin_sector <= end_sector; ++begin_sector) {
+	for (size_t begin_sector = addr / sector_size; begin_sector <= end_sector; ++begin_sector) {
 		/* Erase the current Flash sector */
-		const uint32_t ctrl = (psize * FLASH_CR_PSIZE16) | FLASH_CR_SER | (begin_sector * FLASH_CR_SNB_1);
+		const uint32_t ctrl = stm32h7_flash_cr(sector_size, (psize * FLASH_CR_PSIZE16) | FLASH_CR_SER, begin_sector);
 		target_mem_write32(target, reg_base + FLASH_CR, ctrl);
-		target_mem_write32(target, reg_base + FLASH_CR, ctrl | FLASH_CR_START);
+		const uint32_t ctrl_start = stm32h7_flash_cr(sector_size, ctrl | FLASH_CR_START, begin_sector);
+		target_mem_write32(target, reg_base + FLASH_CR, ctrl_start);
```

Comment:
This results in the same `ctrl` value as before with just ` | FLASH_CR_START`, not sure this change is necessary.

Reply:
This existed in the driver. See RM0433, 4.3.10 Flash Erase Operations. START1 bit may be set *after* SER1 & SNB1 bitfields are set. There is a note that BER1+START1 can be combined for Bank erase, but I'm unsure about Sector erase. This is in all 4 refmans.

Reply:
Yes, we're not saying eliminate the register write - we're saying eliminate the recomputation of `ctrl`. you should still be able to just `target_mem_write32(target, reg_base + FLASH_CR, ctrl | FLASH_CR_START);` and that cheapens the codegen here.

Reply:
As much as I'd like to aim at smaller codegen, START1 is BIT(7) for H74x/H72x, but BIT(5) for H7Bx (and H7RS). Hence the extra call to the shifter to deal with it.

Reply:
Ohhh, we see.. ok, could you please add a comment on the front of this call in each of the places it's made to explain this? We completely missed that detail when reading through this today and that seems like a very important thing not to forget.

It is valid to eliminate the `ctrl_start` variable at least by chaining the `stm32h7_flash_cr()` and `target_mem_write32()` calls though so that might still be a reasonable idea to try and minimise the chances of extra pessimistic codegen here?

Otherwise feel free to disregard this review thread beyond asking for extra comments to guard future readers from making the same mistake.

Reply:
I trust the compiler to optimize away a stack scope const variable (i.e. keep it in gp register only) at -Os. It also helps me with readability (otherwise I'd have to stare at a very long line of code). I added the comment about START1 position.
In the other case (page program) the second invocation of helper is actually unnecessary -- dropped it. First invocation would unflip PSIZE bits, and _PG position is identical.